working/macros/feature-specification.md

```diff
@@ -550,6 +550,116 @@ This does have two interesting and possibly unexpected consequences:
 
 **TODO**: Define the API for adding metadata to existing declarations (#1931).
 
+## Aspect Macros
+
+**NOTE**: This feature is a work in progress, and has many TODOs.
+
+An AspectMacro is used by another macro, to generate shared helper code which
+will not be duplicated across every library where that helper was needed.
+
+Concretely, let's consider the following example:
+
+```dart
+@fromJsonExtension
+class A {}
+
+@fromJsonExtension
+class B {
+  final A a;
+
+  B(this.a);
+}
+
+// Generated extension from the macro on A
+extension AFromJson on Map<String, Object?> {
+  A toA();
+}
+
+// Generated extension from the macro on B
+extension BFromJson on Map<String, Object?> {
+  A toA(); // This also needs to generate a helper for converting `A`!
+
+  B toB();
+}
+```
+
+Both of these macro applications need to generate the same shared code for
+converting a Map into an instance of `A` (since `B` has a field of that type),
+but they can't actually _see_ what each other has generated.
+
+In the worst case, this would result in them both outputting conflicting
+members, and in the best case they would end up generating duplicate code,
+causing unnecessary code bloat.
+
+Aspect Macros exist to solve this problem, by allowing macros to share these
+implementations. This is done by associating a single, canonical declaration
```

Comment:
This solves deduplication when two macro applications would generate the exact same code. But it doesn't solve wanting to *coalesce* information where you want multiple macro applications to build up one single data structure, like a global injector map for DI.

Is that a use case you want to think about for this?

Reply:
Hmm, interesting question. A similar pattern could probably be used to do that (bazel aspects which this is inspired by do allow you to propagate information and merge it together).

I will have a think on that, and add a TODO section.